boot/bootutil/src/loader.c

```diff
@@ -793,58 +855,122 @@ boot_copy_sector(int flash_area_id_src, int flash_area_id_dst,
 static void
 boot_swap_sectors(int idx, uint32_t sz, struct boot_status *bs)
 {
+    const struct flash_area *fap;
     uint32_t copy_sz;
+    uint32_t trailer_sz;
     uint32_t img_off;
+    uint32_t scratch_trailer_off;
+    struct boot_swap_state swap_state;
     int rc;
 
     /* Calculate offset from start of image area. */
     img_off = boot_data.imgs[0].sectors[idx].fa_off -
               boot_data.imgs[0].sectors[0].fa_off;
 
+    copy_sz = sz;
+
+    trailer_sz = boot_slots_trailer_sz(boot_data.write_sz);
+    if (boot_data.imgs[0].sectors[idx].fa_off + sz >
```

Comment:
`sz` in this function is always a size that is corresponds to the full number of bytes of the amount of sectors to copy (never less than a sector size). You're right that this checks against the first address of the last sector and it was done on purpose. What I wanted to know is if we're copying the last sector. If we are, than I discount the trailer, but of course, this is based on the assumption that `sz` will still be the whole last sector. It also breaks if the fw would occupy more than what trailer size leaves for it but that needs to be handled elsewhere.